qa/test-harness/testharness.py

```diff
@@ -0,0 +1,159 @@
+                return requests.post(ode_upload_url, files={'name':'file', 'file':file}, timeout=2)
+        elif self.upload_format == 'BODY':
+            with open(filepath, 'rb') as file:
+                return requests.post(ode_upload_url,
+                    headers={'Content-type': 'application/json', 'Accept': 'application/json'},
+                    data=open(filepath, 'rb'),
+                    timeout=2)
+        else:
+            raise TestHarnessException("Failed to parse configuration section '%s', unknown UploadFormat '%s', expected FILE or BODY." % (self.test_name, self.upload_format))
+
+    def _listen_to_kafka_topics(self, msg_queue, *topics):
+        consumer=KafkaConsumer(*topics, bootstrap_servers=DOCKER_HOST_IP+':'+KAFKA_PORT, consumer_timeout_ms=KAFKA_CONSUMER_TIMEOUT)
+        for msg in consumer:
+            msg_queue.put(str(msg.value, 'utf-8'))
+
+    def run(self):
+        msg_queue = queue.Queue()
+        print("[INFO] Creating Kafka consumer listenting on topic(s): %s ..." % self.list_of_kafka_topics)
+        kafkaListenerThread=threading.Thread(target=self._listen_to_kafka_topics, args=(msg_queue, *self.list_of_kafka_topics))
+        kafkaListenerThread.start()
+        time.sleep(3)
+        print("[INFO] Kafka consumer preparation complete.")
+
+        # Upload the test file with known data to the ODE
+        print("[INFO] Uploading test file to ODE...")
+        try:
+            upload_response = self._upload_file(self.data_file_path, self.upload_url)
+            print("[INFO] Upload response received: %s %s" % (upload_response.status_code, upload_response.text))
+            if upload_response.status_code == 200:
+                print("[INFO] Test file uploaded successfully.")
+            else:
+                print("[ERROR] Aborting test routine! Test file (%s) failed to upload to (%s), response code %d" % (self.data_file_path, self.upload_url, upload_response.status_code))
+                return
+        except (requests.exceptions.ConnectionError, requests.exceptions.ConnectTimeout) as e:
+            print("[ERROR] Aborting test routine! Test file upload failed (unable to reach ODE). Error: '%s'" % str(e))
+            return
+
+        # Wait for as many messages as possible to be collected
+        print("[INFO] Waiting for all messages to be received...")
+        kafkaListenerThread.join()
+        print("[INFO] Received %d messages from Kafka consumer." % msg_queue.qsize())
+
+        if msg_queue.qsize() == 0:
+            raise TestHarnessException("[ERROR] Aborting test routine! Received no messages from the Kafka consumer.")
+
+        # After all messages were received, log them to a file
+        self.validation_results = self.validator.validate_queue(msg_queue)
+        self.bundle_id = self.validation_results[0].record['metadata']['serialId']['bundleId']
+        self.serial_number = self.validation_results[0].record['metadata']['serialId']['serialNumber']
+
```

Comment:
Save the `['metadata']['serialId']['streamId']` here in the Iteration object so it can be used later by the bundle ID check and not having to use the config file hack.

Reply:
Done.